<?xml version='1.0' encoding='UTF-8'?>
<sdf version="1.6">
  <model name="orca4_ign">
    <pose>0 0 0 0 0 0</pose>

    <!-- BlueROV2 thruster reference: http://www.ardusub.com/introduction/features.html -->
    <!-- fluid_density: 1000 -->
    <!-- displaced_mass: 10.05 -->

    <link name="base_link">
      <inertial>
        <pose>0 0 0.011 0 0 0</pose>
        <mass>11.035</mass>
        <inertia>
          <ixx>0.09873097998042396</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.17756847998042397</iyy>
          <iyz>0</iyz>
          <izz>0.2692441666666667</izz>
        </inertia>
      </inertial>

      <visual name="base_link_visual">
        <geometry>
          <mesh>
            <uri>model://auv_slam/models/orca4_ign/meshes/bluerov2.dae</uri>
            <scale>0.025 0.025 0.025</scale>
          </mesh>
        </geometry>
      </visual>

      <!-- Collision volume * fluid_density == displaced mass, used by BuoyancyPlugin -->
      <collision name="base_link_collision">
        <pose>0 0 0.06 0 0 0</pose>
        <geometry>
          <box>
            <size>0.457 0.338 0.06506286173008947</size>
          </box>
        </geometry>
      </collision>

    </link>

    <!-- Hydrodynamics plugin -->
    <plugin
        filename="ignition-gazebo-hydrodynamics-system"
        name="ignition::gazebo::systems::Hydrodynamics">
      <link_name>base_link</link_name>
      <water_density>1000</water_density>
      <!-- Added mass: -->
      <xDotU>0</xDotU>
      <yDotV>0</yDotV>
      <zDotW>0</zDotW>
      <kDotP>0</kDotP>
      <mDotQ>0</mDotQ>
      <nDotR>0</nDotR>
      <!-- First order stability derivative: -->
      <xU>0</xU>
      <yV>0</yV>
      <zW>0</zW>
      <kP>0</kP>
      <mQ>0</mQ>
      <nR>0</nR>
      <!-- Second order stability derivative: -->
      <xUabsU>-33.800000000000004</xUabsU>
      <yVabsV>-54.26875</yVabsV>
      <zWabsW>-73.37135</zWabsW>
      <kPabsP>-4.0</kPabsP>
      <mQabsQ>-4.0</mQabsQ>
      <nRabsR>-4.0</nRabsR>
    </plugin>

    <link name="thruster1">
      <pose>0.14 -0.092 0 -1.5707963267948966 1.5707963267948966 -0.7853981633974483</pose>
      <visual name="thruster_prop_visual">
        <pose>0 0 0 1.5707963267948966 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://auv_slam/models/orca4_ign/meshes/t200-cw-prop.dae</uri>
          </mesh>
        </geometry>
      </visual>
      <inertial>
        <mass>0.002</mass>
        <inertia>
          <ixx>0.001</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.001</iyy>
          <iyz>0</iyz>
          <izz>0.001</izz>
        </inertia>
      </inertial>
    </link>

    <link name="thruster2">
      <pose>0.14 0.092 0 -1.5707963267948966 1.5707963267948966 -2.356194490192345</pose>
      <visual name="thruster_prop_visual">
        <pose>0 0 0 -1.5707963267948966 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://auv_slam/models/orca4_ign/meshes/t200-ccw-prop.dae</uri>
          </mesh>
        </geometry>
      </visual>
      <inertial>
        <mass>0.002</mass>
        <inertia>
          <ixx>0.001</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.001</iyy>
          <iyz>0</iyz>
          <izz>0.001</izz>
        </inertia>
      </inertial>
    </link>

    <link name="thruster3">
      <pose>-0.14 -0.092 0 -1.5707963267948966 1.5707963267948966 0.7853981633974483</pose>
      <visual name="thruster_prop_visual">
        <pose>0 0 0 -1.5707963267948966 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://auv_slam/models/orca4_ign/meshes/t200-ccw-prop.dae</uri>
          </mesh>
        </geometry>
      </visual>
      <inertial>
        <mass>0.002</mass>
        <inertia>
          <ixx>0.001</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.001</iyy>
          <iyz>0</iyz>
          <izz>0.001</izz>
        </inertia>
      </inertial>
    </link>

    <link name="thruster4">
      <pose>-0.14 0.092 0 -1.5707963267948966 1.5707963267948966 2.356194490192345</pose>
      <visual name="thruster_prop_visual">
        <pose>0 0 0 1.5707963267948966 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://auv_slam/models/orca4_ign/meshes/t200-cw-prop.dae</uri>
          </mesh>
        </geometry>
      </visual>
      <inertial>
        <mass>0.002</mass>
        <inertia>
          <ixx>0.001</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.001</iyy>
          <iyz>0</iyz>
          <izz>0.001</izz>
        </inertia>
      </inertial>
    </link>

    <link name="thruster5">
      <pose>0 -0.109 0.077 0 0 0</pose>
      <visual name="thruster_prop_visual">
        <pose>0 0 0 -1.5707963267948966 0 1.5707963267948966</pose>
        <geometry>
          <mesh>
            <uri>model://auv_slam/models/orca4_ign/meshes/t200-ccw-prop.dae</uri>
          </mesh>
        </geometry>
      </visual>
      <inertial>
        <mass>0.002</mass>
        <inertia>
          <ixx>0.001</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.001</iyy>
          <iyz>0</iyz>
          <izz>0.001</izz>
        </inertia>
      </inertial>
    </link>

    <link name="thruster6">
      <pose>0 0.109 0.077 0 0 0</pose>
      <visual name="thruster_prop_visual">
        <pose>0 0 0 1.5707963267948966 0 1.5707963267948966</pose>
        <geometry>
          <mesh>
            <uri>model://auv_slam/models/orca4_ign/meshes/t200-cw-prop.dae</uri>
          </mesh>
        </geometry>
      </visual>
      <inertial>
        <mass>0.002</mass>
        <inertia>
          <ixx>0.001</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.001</iyy>
          <iyz>0</iyz>
          <izz>0.001</izz>
        </inertia>
      </inertial>
    </link>

    <joint name="thruster1_joint" type="revolute">
      <pose relative_to="base_link">0.14 -0.092 0 -1.5707963267948966 1.5707963267948966 -0.7853981633974483</pose>
      <parent>base_link</parent>
      <child>thruster1</child>
      <axis>
          <xyz>0 0 -1</xyz>
          <limit>
            <lower>-inf</lower>
            <upper>inf</upper>
          </limit>
      </axis>
    </joint>

    <joint name="thruster2_joint" type="revolute">
      <pose relative_to="base_link">0.14 0.092 0 -1.5707963267948966 1.5707963267948966 -2.356194490192345</pose>
      <parent>base_link</parent>
      <child>thruster2</child>
      <axis>
        <xyz>0 0 -1</xyz>
        <limit>
            <lower>-inf</lower>
            <upper>inf</upper>
        </limit>
      </axis>
    </joint>

    <joint name="thruster3_joint" type="revolute">
      <pose relative_to="base_link">-0.14 -0.092 0 -1.5707963267948966 1.5707963267948966 0.7853981633974483</pose>
      <parent>base_link</parent>
      <child>thruster3</child>
      <axis>
        <xyz>0 0 -1</xyz>
        <limit>
            <lower>-inf</lower>
            <upper>inf</upper>
        </limit>
      </axis>
    </joint>

    <joint name="thruster4_joint" type="revolute">
      <pose relative_to="base_link">-0.14 0.092 0 -1.5707963267948966 1.5707963267948966 2.356194490192345</pose>
      <parent>base_link</parent>
      <child>thruster4</child>
      <axis>
        <xyz>0 0 -1</xyz>
        <limit>
            <lower>-inf</lower>
            <upper>inf</upper>
        </limit>
      </axis>
    </joint>

    <joint name="thruster5_joint" type="revolute">
      <pose relative_to="base_link">0 -0.109 0.077 0 0 0</pose>
      <parent>base_link</parent>
      <child>thruster5</child>
      <axis>
        <xyz>0 0 -1</xyz>
        <limit>
            <lower>-inf</lower>
            <upper>inf</upper>
        </limit>
      </axis>
    </joint>

    <joint name="thruster6_joint" type="revolute">
      <pose relative_to="base_link">0 0.109 0.077 0 0 0</pose>
      <parent>base_link</parent>
      <child>thruster6</child>
      <axis>
        <xyz>0 0 -1</xyz>
        <limit>
            <lower>-inf</lower>
            <upper>inf</upper>
        </limit>
      </axis>
    </joint>

    <plugin
        filename="ignition-gazebo-thruster-system"
        name="ignition::gazebo::systems::Thruster">
      <namespace>orca4_ign</namespace>
      <joint_name>thruster1_joint</joint_name>
      <thrust_coefficient>0.5</thrust_coefficient>
      <fluid_density>1000</fluid_density>
      <propeller_diameter>0.1</propeller_diameter>
    </plugin>

    <plugin
        filename="ignition-gazebo-thruster-system"
        name="ignition::gazebo::systems::Thruster">
      <namespace>orca4_ign</namespace>
      <joint_name>thruster2_joint</joint_name>
      <thrust_coefficient>0.5</thrust_coefficient>
      <fluid_density>1000</fluid_density>
      <propeller_diameter>0.1</propeller_diameter>
    </plugin>

    <plugin
        filename="ignition-gazebo-thruster-system"
        name="ignition::gazebo::systems::Thruster">
      <namespace>orca4_ign</namespace>
      <joint_name>thruster3_joint</joint_name>
      <!-- Reverse spin to balance torque -->
      <thrust_coefficient>-0.5</thrust_coefficient>
      <fluid_density>1000</fluid_density>
      <propeller_diameter>0.1</propeller_diameter>
    </plugin>

    <plugin
        filename="ignition-gazebo-thruster-system"
        name="ignition::gazebo::systems::Thruster">
      <namespace>orca4_ign</namespace>
      <joint_name>thruster4_joint</joint_name>
      <!-- Reverse spin to balance torque -->
      <thrust_coefficient>-0.5</thrust_coefficient>
      <fluid_density>1000</fluid_density>
      <propeller_diameter>0.1</propeller_diameter>
    </plugin>

    <plugin
        filename="ignition-gazebo-thruster-system"
        name="ignition::gazebo::systems::Thruster">
      <namespace>orca4_ign</namespace>
      <joint_name>thruster5_joint</joint_name>
      <thrust_coefficient>0.5</thrust_coefficient>
      <fluid_density>1000</fluid_density>
      <propeller_diameter>0.1</propeller_diameter>
    </plugin>

    <plugin
        filename="ignition-gazebo-thruster-system"
        name="ignition::gazebo::systems::Thruster">
      <namespace>orca4_ign</namespace>
      <joint_name>thruster6_joint</joint_name>
      <!-- Reverse spin to balance torque -->
      <thrust_coefficient>-0.5</thrust_coefficient>
      <fluid_density>1000</fluid_density>
      <propeller_diameter>0.1</propeller_diameter>
    </plugin>

    <link name="camera_forward_link">
      <pose>0.15 0.0 -0.0675 0 0 0</pose>
      <inertial>
        <mass>0.001</mass>
        <inertia>
          <ixx>0.001</ixx> <iyy>0.001</iyy> <izz>0.001</izz>
        </inertia>
      </inertial>
      <sensor name="camera_forward" type="camera">
        <camera>
          <horizontal_fov>1.4</horizontal_fov>
          <image>
            <width>800</width> <height>600</height>
          </image>
          <clip>
            <near>0.1</near> <far>50.0</far>
          </clip>
        </camera>
        <always_on>1</always_on>
        <update_rate>10</update_rate>
        <visualize>true</visualize>
        <topic>/camera_forward/image_raw</topic>
        <camera_info_topic>/camera_forward/camera_info</camera_info_topic>
      </sensor>
    </link>

    <joint name="camera_forward_joint" type="fixed">
      <parent>base_link</parent>
      <child>camera_forward_link</child>
    </joint>

    <link name="camera_down_link">
      <pose>0.1 0 -0.1 0 -1.570796 0</pose> 
      <inertial>
        <mass>0.001</mass>
        <inertia>
          <ixx>0.001</ixx> <iyy>0.001</iyy> <izz>0.001</izz>
        </inertia>
      </inertial>
      <sensor name="camera_down" type="camera">
        <camera>
          <horizontal_fov>1.4</horizontal_fov>
          <image>
            <width>640</width> <height>480</height>
          </image>
          <clip>
            <near>0.1</near> <far>20.0</far>
          </clip>
        </camera>
        <always_on>1</always_on>
        <update_rate>5</update_rate>
        <visualize>true</visualize>
        <topic>/camera_down/image_raw</topic>
        <camera_info_topic>/camera_down/camera_info</camera_info_topic>
      </sensor>
    </link>
    
    <joint name="camera_down_joint" type="fixed">
      <parent>base_link</parent>
      <child>camera_down_link</child>
    </joint>

    <!-- Publish the ground-truth pose for rviz -->
    <plugin
        filename="ignition-gazebo-odometry-publisher-system"
        name="ignition::gazebo::systems::OdometryPublisher">
      <odom_frame>map</odom_frame>
      <robot_base_frame>orca4_ign</robot_base_frame>
      <dimensions>3</dimensions>
    </plugin>

  

    <link name="imu_link">
      <visual name="imu_visual">
        <geometry>
          <box>
          <size>0.1 0.1 0.1</size>
          </box>
        </geometry>
      </visual>

      <collision name="imu_collision">
        <geometry>
          <box>
          <size>0.1 0.1 0.1</size>
          </box>
        </geometry>
      </collision>

      <inertial>
        <origin xyz="0 0 0" rpy="3.141592653589793 0 0"/>
        <mass>0.001</mass>
        <inertia>
          <ixx>0.001</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.001</iyy>
          <iyz>0</iyz>
          <izz>0.001</izz>
        </inertia>
      </inertial>
    

      <sensor name="imu_sensor" type="imu">
        <ignition_frame_id>imu_link</ignition_frame_id>
        <plugin filename="ignition-gazebo-imu-system" name="ignition::gazebo::systems::Imu">
        </plugin>

        <topic>/demo/imu</topic>
        <always_on>true</always_on>
        <update_rate>100</update_rate>
        <visualize>true</visualize>
        <enable_metrics>true</enable_metrics>
        <imu>
          <angular_velocity>
            <x>
              <noise type="gaussian">
                <mean>0.0</mean>
                <stddev>2e-4</stddev>
                <bias_mean>0.0000075</bias_mean>
                <bias_stddev>0.0000008</bias_stddev>
              </noise>
            </x>
            <y>
              <noise type="gaussian">
                <mean>0.0</mean>
                <stddev>2e-4</stddev>
                <bias_mean>0.0000075</bias_mean>
                <bias_stddev>0.0000008</bias_stddev>
              </noise>
            </y>
            <z>
              <noise type="gaussian">
                <mean>0.0</mean>
                <stddev>2e-4</stddev>
                <bias_mean>0.0000075</bias_mean>
                <bias_stddev>0.0000008</bias_stddev>
              </noise>
            </z>
          </angular_velocity>
          <linear_acceleration>
            <x>
              <noise type="gaussian">
                <mean>0.0</mean>
                <stddev>1.7e-2</stddev>
                <bias_mean>0.1</bias_mean>
                <bias_stddev>0.001</bias_stddev>
              </noise>
            </x>
            <y>
              <noise type="gaussian">
                <mean>0.0</mean>
                <stddev>1.7e-2</stddev>
                <bias_mean>0.1</bias_mean>
                <bias_stddev>0.001</bias_stddev>
              </noise>
            </y>
            <z>
              <noise type="gaussian">
                <mean>0.0</mean>
                <stddev>1.7e-2</stddev>
                <bias_mean>0.1</bias_mean>
                <bias_stddev>0.001</bias_stddev>
              </noise>
            </z>
          </linear_acceleration>
        </imu>
      </sensor>
    </link>

  <joint name="imu_joint" type="fixed">
    <parent>base_link</parent>
    <child>imu_link</child>
    <pose relative_to="base_link">0 0 0.01 0 0 0</pose>
  </joint>

  <!--

    <plugin
      filename="ignition-gazebo-joint-state-publisher-system"
      name="ignition::gazebo::systems::JointStatePublisher">
      <topic>joint_states</topic>
    </plugin>

    

    <joint name='dvl_joint' type='fixed'>
      <parent>base_link</parent>
      <child>dvl_link</child>
      <pose relative_to="base_link">0.0 0.0 0.01 0 0 0</pose>
    </joint>

    <link name='dvl_link'>
      <pose relative_to="dvl_joint"/>
      <visual name="dvl_link_visual">
        <geometry>
          <box><size>
            0.1 0.1 0.1
          </size></box>
        </geometry>
      </visual>

      <collision name="dvl_link_collision">
        <geometry>
          <box><size>
            0.1 0.1 0.1
          </size></box>
        </geometry>
      </collision>

      <inertial>
          <pose>0 0 0 3.141592653589793 0 0</pose>
          <inertial>
            <mass>0.001</mass>
            <inertia>
              <ixx>0.001</ixx>
              <ixy>0</ixy>
              <ixz>0</ixz>
              <iyy>0.001</iyy>
              <iyz>0</iyz>
              <izz>0.001</izz>
            </inertia>
          </inertial>
      </inertial>

      <sensor name="dvl_sensor" type="dvl">
        <always_on>true</always_on>
        <update_rate>100</update_rate>
        <visualize>true</visualize>
        <topic>demo/odom</topic>
        <gz_frame_id>dvl_link</gz_frame_id>
      </sensor>
    </link>

    -->
  </model>
</sdf>